compiler/rustc_session/src/options.rs

```diff
@@ -1463,8 +1478,11 @@ options! {
         "output statistics about monomorphization collection"),
     dump_mono_stats_format: DumpMonoStatsFormat = (DumpMonoStatsFormat::Markdown, parse_dump_mono_stats, [UNTRACKED],
         "the format to use for -Z dump-mono-stats (`markdown` (default) or `json`)"),
-    dump_solver_proof_tree: bool = (false, parse_bool, [UNTRACKED],
-        "dump a proof tree for every goal evaluated by the new trait solver"),
+    dump_solver_proof_tree: SolverProofTreeCondition = (SolverProofTreeCondition::OnRequest, parse_solver_proof_tree_condition, [UNTRACKED],
+        "dump a proof tree for every goal evaluated by the new trait solver. If the flag is specified without any options after it
+        then it defaults to `always`. If the flag is not specified at all it defaults to `on-request`."),
+    dump_solver_proof_tree_use_cache: Option<bool> = (None, parse_opt_bool, [UNTRACKED],
+        "determines whether proof tree generation uses the global cache"),
```

Comment:
```suggestion
        "determines whether dumping proof trees should disable the global cache"),
```
this flag should not affect general proof tree generation, only the proof trees which should get dumped

Reply:
that's.... not how the code works. the flag _does_ cause us to use or not use the global cache during proof tree creation. It's not a step done when we dump the proof tree to edit all the "GLOBAL CACHE HIT" to actual proof trees. Implementing it that way sounds trickier than what was done in this PR

Reply:
from the description:
> This flag currently would affect misc uses of GenerateProofTree::Yes which will be added in the future for things like diagnostics logic and rustdoc's auto_trait file. We can fix this when we start using proof tree generation for those use cases if it's desirable.

When `-Zdump-solver-proof-tree-use-cache=no/yes` is set it acts on every call to `evaluate_goal` with `GenerateProofTree::Yes` and also to all proof trees generated by `-Zdump-solver-proof-tree=always`

Reply:
my thought was that setting this to `true` does/will not prevent diagnostics and auto trait computation from disabling the cache, so this only affects trait solving when we construct a tree for "proof tree dumping"

Reply:
ideally that's how it would work but right now it wouldn't if you implemented auto_trait/diagnostics by callign `evalaute_goal` with `GenerateProofTree::Yes(UseGlobalCache::No)`.

Reply:
yes, but that would completely ignore this compile flag